examples/tv-app/tv-common/clusters/messages/MessagesManager.h

```diff
@@ -21,6 +21,115 @@
 
 #include <iostream>
 #include <list>
+#include <vector>
+
+struct CachedMessageOption
+{
+    constexpr static size_t kResponseIdMin  = 1;
+    constexpr static size_t kLabelMaxLength = 32;
+
+    CachedMessageOption(uint32_t id, std::string label) :
+        mLabel(label), mOption{ chip::MakeOptional(id), chip::MakeOptional(chip::CharSpan::fromCharString(mLabel.c_str())) }
+    {}
+
+    CachedMessageOption(const CachedMessageOption & option) :
+        mLabel(option.mLabel),
+        mOption{ option.mOption.messageResponseID, chip::MakeOptional(chip::CharSpan::fromCharString(mLabel.c_str())) } {};
+
+    CachedMessageOption & operator=(const CachedMessageOption & option) { return *this; };
+
+    chip::app::Clusters::Messages::MessageResponseOption GetMessageOption() { return mOption; }
+
+    ~CachedMessageOption() {}
+
+protected:
+    std::string mLabel;
+    chip::app::Clusters::Messages::MessageResponseOption mOption;
+};
+
+struct CachedMessage
+{
+    constexpr static size_t kMessageIdLength = 16;
+    constexpr static size_t kMaxOptionCount  = 4;
+
+    chip::ByteSpan mMessageId;
+
+    CachedMessage(const CachedMessage & message) :
+        mPriority(message.mPriority), mMessageControl(message.mMessageControl), mStartTime(message.mStartTime),
+        mDuration(message.mDuration), mMessageText(message.mMessageText), mOptions(message.mOptions)
+    {
+        memcpy(mMessageIdBuffer, message.mMessageIdBuffer, kMessageIdLength);
+        mMessageId = chip::ByteSpan(mMessageIdBuffer, kMessageIdLength);
+
+        for (CachedMessageOption & entry : mOptions)
+        {
+            mResponseOptions.push_back(entry.GetMessageOption());
+        }
+    };
+
+    CachedMessage & operator=(const CachedMessage & message) { return *this; };
+
+    CachedMessage(const chip::ByteSpan & messageId, const chip::app::Clusters::Messages::MessagePriorityEnum & priority,
+                  const chip::BitMask<chip::app::Clusters::Messages::MessageControlBitmap> & messageControl,
+                  const chip::app::DataModel::Nullable<uint32_t> & startTime,
+                  const chip::app::DataModel::Nullable<uint16_t> & duration, std::string messageText) :
+        mPriority(priority),
+        mMessageControl(messageControl), mStartTime(startTime), mDuration(duration), mMessageText(messageText)
+    {
+        if (messageId.size() != kMessageIdLength)
+        {
+            ChipLogProgress(Zcl, "CachedMessage bad message id size, ignoring");
```

Comment:
This check needs to be way before this and result in errors to client.

Reply:
moved checks to root cluster (outside delegate)